4548 Modify event avro schema spec to use dynamic lookup

brendan-git-va opened the pull request:
## Summary

- This work is behind a feature toggle (flipper): YES
- Modified the spec for the ST event payload validation to no longer use hard coded references to the spec.  The changes are based on a previous ticket, [4481](https://github.com/department-of-veterans-affairs/VES/issues/4481) to allow the dynamic schema reference.
- The work was done for the Submission Traceability team and is part of the ongoing project to track submissions.

## Related issue(s)

- [Modify event Avro schema to use dynamic look up ](https://github.com/department-of-veterans-affairs/VES/issues/4548)

## Testing done

- The spec file previously referenced the schema file through hard coded logic.  This has been changed to a reference and uses the VCR gem for testing
- The tests now use the public method `produce` instead of the private method `validate_payload`
- No new code was added outside of the spec file itself

## What areas of the site does it impact?
 - The changes are only to the ongoing work on the Kafka producer and will not impact anything outside that scope

## Acceptance criteria

- [x]  I fixed|updated|added unit tests and integration tests for each feature (if applicable).
- [x]  No error nor warning in the console.
- [ ]  Events are being sent to the appropriate logging solution
- [ ]  Documentation has been updated (link to documentation)
- [x]  No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs
- [ ]  Feature/bug has a monitor built into Datadog (if applicable)
- [ ]  If app impacted requires authentication, did you login to a local build and verify all authenticated routes work as expected
- [ ]  I added a screenshot of the developed feature

madebydna commented on spec/lib/kafka/avro_producer_payload_spec.rb:
```diff
@@ -24,127 +21,182 @@
+    PAYLOAD
+  end
 
   describe '#validate_payload!' do
```
Could you rename this describe block to not directly reference the private method `#validate_payload!`? I see that you're no longer testing that method but the public `.produce` method instead, which is great. You can just call the describe block something like "validate payload" instead. 🙂

madebydna reviewed (approved):
Looks good! Just a small change request on naming the `describe` block in the specs.

brendan-git-va commented on spec/support/vcr_cassettes/kafka/topics.yml:
```diff
@@ -183,5 +183,127 @@ http_interactions:
+                  "subject": "topic-3-value",
+                  "version": 1,
+                  "id": 5,
+                  "schema": "{\"type\":\"record\",\"name\":\"SubmissionTrace\",\"namespace\":\"gov.va.eventbus.test.data\",\"fields\":[{\"name\":\"priorId\",\"type\":[\"null\",\"string\"]},{\"name\":\"currentId\",\"type\":\"string\"},{\"name\":\"nextId\",\"type\":[\"null\",\"string\"]},{\"name\":\"icn\",\"type\":[\"null\",\"string\"]},{\"name\":\"vasiId\",\"type\":\"string\"},{\"name\":\"systemNa ...
```
Done

brendan-git-va commented on spec/support/vcr_cassettes/kafka/topics.yml:
```diff
@@ -183,5 +183,127 @@ http_interactions:
+  recorded_at: Tue, 10 Aug 2021 15:55:41 GMT  
+- request:
+    method: get
+    uri: http://localhost:8081/subjects/topic-4-value/versions/latest
```
This is from when I was testing the schema the same way it was tested in avro_producer_spec.  It is no longer needed so I updated topic.yml to only have topics 1, 2 and 3

brendan-git-va commented on spec/lib/kafka/avro_producer_payload_spec.rb:
```diff
@@ -24,127 +21,170 @@
 
-  describe '#validate_payload!' do
+  before do
+    allow(Flipper).to receive(:enabled?).with(:kafka_producer_fetch_schema_dynamically).and_return(true)
```
I think this got mixed in with the schema errors I was encountering elsewhere.  I tried removing it after fixing those and you're right.  It is not needed

brendan-git-va commented on spec/lib/kafka/avro_producer_payload_spec.rb:
```diff
@@ -24,127 +21,170 @@
   end
   let(:invalid_payload_format) { 'invalid' }
+  let(:topic3_payload_value) do
+    "\x00\x00\x00\x00\x05\x00\n12345\x00\x02\x12ICN123456\x12VASI98765\x00\x00\x00\x00\x00\x00(2024-03-04T12:00:00Z\x00"
```
This was a couple issues that were causing the encoding to change depending on how I setup topic3_payload_value to avoid lint errors.  
The message came back in ASCII-8BIT encoding, but topic3_payload_value was in UTF-8.

I added ".b" to the end of the string to set the encoding to ASCII-8BIT and that fixed the issue

brendan-git-va commented:
Fixed the issues mentioned along with several suggestions from Event Bus

- Changed the namespace to gov.va.submissiontrace.form.data
- Changed the names for types to PascalCase (SystemName and SumissionName)
- Added doc property to each field, based on this [doc](https://github.com/department-of-veterans-affairs/VES/blob/master/cx-measurement/submission-traceability/documentation/event-payload.md)

christineyost commented:
The schema looks good! Thanks for all the updates, especially the docs 🎉

madebydna reviewed (approved):
Looks good! 👍
